### ISSUE TITLE
JDK18 java/nio/channels/FileChannel/LargeGatheringWrite.java - OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse-openj9/openj9#14503]
Failure link
------------

From an internal build `job/Test_openjdk18_j9_extended.openjdk_ppc64_aix/3/`(`paix907`):
```
openjdk version "18" 2022-03-22
IBM Semeru Runtime Open Edition 18.0.0.0-m1 (build 18+35)
Eclipse OpenJ9 VM 18.0.0.0-m1 (build openj9-0.31.0-m1, JRE 18 AIX ppc64-64-Bit Compressed References 20220210_28 (JIT enabled, AOT enabled)
OpenJ9   - 700cde7b8
OMR      - bb5d44b8c
JCL      - cdd4205c649 based on jdk-18+35)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=upstream&amp;TARGET=jdk_nio_0&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=0&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=adoptium%3Amaster&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=ppc64_aix&amp;GENERATE_JOBS=false&amp;KEEP_REPORTDIR=true&amp;PERSONAL_BUILD=false&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;LABEL=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-aix-ppc64-openj9%2F28%2Fibm-semeru-open-jdk_ppc64_aix_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-aix-ppc64-openj9%2F28%2Fibm-semeru-open-jre_ppc64_aix_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-aix-ppc64-openj9%2F28%2Fibm-semeru-open-debugimage_ppc64_aix_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-aix-ppc64-openj9%2F28%2Fibm-semeru-open-testimage_ppc64_aix_18_35_openj9-0.31.0-m1.tar.gz&amp;BUILD_IDENTIFIER=&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=true&amp;NON_AQA_TEST_REPOS=&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=&amp;JDK_VERSION=18&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=https%3A%2F%2Fgithub.com%2Fibmruntimes%2Fopenj9-openjdk-jdk18&amp;RELEASE_TAG=v0.31.0-release&amp;OPENJ9_BRANCH=v0.31.0-release&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;PLATFORM_AND_MACHINE=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=v0.31.0-release&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=&amp;ARTIFACTORY_ROOT_DIR=&amp;UPSTREAM_TEST_JOB_NUMBER=&amp;DOCKERIMAGE_TAG=&amp;JDK_IMPL=openj9&amp;TEST_TIME=&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;SLACK_CHANNEL=&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=4e18ffe7-b1b1-4272-9979-99769b68bcc2&amp;ARCHIVE_TEST_RESULTS=false&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=http%3A%2F%2Ftrss1.fyre.ibm.com&amp;BUILD_LIST=openjdk&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TIME_LIMIT=20&amp;JVM_OPTIONS=&amp;PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2022-02-11T07:42:13.549Z] variation: Mode150
[2022-02-11T07:42:13.549Z] JVM_OPTIONS:  -XX:+UseCompressedOops 

[2022-02-11T08:08:10.204Z] TEST: java/nio/channels/FileChannel/LargeGatheringWrite.java

[2022-02-11T08:08:10.205Z] STDOUT:
[2022-02-11T08:08:10.205Z] Seed from RandomFactory = -8247103834647789495L
[2022-02-11T08:08:10.205Z] Writing 5368709122 bytes of data...
[2022-02-11T08:08:10.205Z] STDERR:
[2022-02-11T08:08:10.205Z] java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
[2022-02-11T08:08:10.205Z] 	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
[2022-02-11T08:08:10.205Z] 	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123)
[2022-02-11T08:08:10.205Z] 	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:332)
[2022-02-11T08:08:10.205Z] 	at java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:243)
[2022-02-11T08:08:10.205Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:209)
[2022-02-11T08:08:10.205Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:165)
[2022-02-11T08:08:10.205Z] 	at java.base/sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:317)
[2022-02-11T08:08:10.205Z] 	at java.base/java.nio.channels.FileChannel.write(FileChannel.java:436)
[2022-02-11T08:08:10.205Z] 	at LargeGatheringWrite.main(LargeGatheringWrite.java:91)
[2022-02-11T08:08:10.205Z] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
[2022-02-11T08:08:10.205Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
[2022-02-11T08:08:10.205Z] 	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
[2022-02-11T08:08:10.205Z] 	at java.base/java.lang.Thread.run(Thread.java:889)
[2022-02-11T08:08:10.205Z] 
[2022-02-11T08:08:10.205Z] JavaTest Message: Test threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)

[2022-02-11T08:08:10.206Z] TEST RESULT: Failed. Execution failed: `main' threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
```

`job/Test_openjdk18_j9_extended.openjdk_s390x_linux/5/`(`ub18s390xrt-1-9`)
```
[2022-02-11T07:08:46.817Z] variation: Mode650
[2022-02-11T07:08:46.817Z] JVM_OPTIONS:  -XX:-UseCompressedOops 

[2022-02-11T07:14:29.129Z] TEST: java/nio/channels/FileChannel/LargeGatheringWrite.java

[2022-02-11T07:14:29.131Z] TEST RESULT: Failed. Unexpected exit from test [exit code: 137]
[2022-02-11T07:14:29.131Z] --------------------------------------------------
[2022-02-11T07:34:06.023Z] Test results: passed: 445; failed: 1
[2022-02-11T07:34:23.488Z] Report written to /home/jenkins/workspace/Test_openjdk18_j9_extended.openjdk_s390x_linux/jvmtest/openjdk/report/html/report.html
[2022-02-11T07:34:23.488Z] Results written to /home/jenkins/workspace/Test_openjdk18_j9_extended.openjdk_s390x_linux/aqa-tests/TKG/output_16445604291340/jdk_nio_1/work
[2022-02-11T07:34:23.488Z] Error: Some tests failed or other problems occurred.
[2022-02-11T07:34:23.488Z] 
[2022-02-11T07:34:23.488Z] jdk_nio_1_FAILED
```

`job/Test_openjdk18_j9_extended.openjdk_x86-64_windows/5/`(`win16x64rt-1-2`)
```
[2022-02-11T08:03:13.395Z] variation: Mode150
[2022-02-11T08:03:13.833Z] JVM_OPTIONS:  -XX:+UseCompressedOops 

openjdk version "18" 2022-03-22
IBM Semeru Runtime Open Edition 18.0.0.0-m1 (build 18+35)
Eclipse OpenJ9 VM 18.0.0.0-m1 (build openj9-0.31.0-m1, JRE 18 Windows Server 2016 amd64-64-Bit Compressed References 20220211_28 (JIT enabled, AOT enabled)
OpenJ9   - 700cde7b8
OMR      - bb5d44b8c
JCL      - cdd4205c649 based on jdk-18+35)

[2022-02-11T08:09:00.188Z] TEST: java/nio/channels/FileChannel/LargeGatheringWrite.java

[2022-02-11T08:09:00.190Z] STDERR:
[2022-02-11T08:09:00.190Z] java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
[2022-02-11T08:09:00.190Z] 	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
[2022-02-11T08:09:00.190Z] 	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123)
[2022-02-11T08:09:00.190Z] 	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:332)
[2022-02-11T08:09:00.190Z] 	at java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:243)
[2022-02-11T08:09:00.190Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:209)
[2022-02-11T08:09:00.190Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:165)
[2022-02-11T08:09:00.190Z] 	at java.base/sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:317)
[2022-02-11T08:09:00.190Z] 	at java.base/java.nio.channels.FileChannel.write(FileChannel.java:436)
[2022-02-11T08:09:00.190Z] 	at LargeGatheringWrite.main(LargeGatheringWrite.java:91)
[2022-02-11T08:09:00.190Z] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
[2022-02-11T08:09:00.190Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
[2022-02-11T08:09:00.190Z] 	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
[2022-02-11T08:09:00.190Z] 	at java.base/java.lang.Thread.run(Thread.java:889)
[2022-02-11T08:09:00.190Z] 
[2022-02-11T08:09:00.190Z] JavaTest Message: Test threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)

```

**Comment by tajila** [comment]
Looking at the test, it writes 3Gbs of data, half of the data comes from a heap buffer the other half from a direct buffer.
When attempting to write the contents of the heap buffer `java.base/sun.nio.ch.FileChannelImpl.write` will allocate a temporary direct buffer, this is where we fail with OOM, m guess is that we have reached the memory limit.

**Comment by tajila** [comment]
@JasonFengJ9 Does this fail on all platforms?

**Comment by pshipton** [comment]
I started running a few open grinders on this test (just 1 iteration). It failed on AIX but not alinux or zlinux.

**Comment by pshipton** [comment]
@tajila any progress on this one?

**Comment by JasonFengJ9** [comment]
> Does this fail on all platforms?

Oops missed this, it also failed at Windows - `job/Test_openjdk18_j9_extended.openjdk_x86-64_windows/8/` (`win16x64rt-1-7`)
```
openjdk version "18" 2022-03-22
IBM Semeru Runtime Open Edition 18.0.0.0-m2 (build 18+36)
Eclipse OpenJ9 VM 18.0.0.0-m2 (build openj9-0.31.0-m2, JRE 18 Windows Server 2016 amd64-64-Bit Compressed References 20220224_35 (JIT enabled, AOT enabled)
OpenJ9   - 6c2d82c95
OMR      - 3792851d0
JCL      - 303814df05f based on jdk-18+36)

[2022-02-24T05:21:01.277Z] TEST: java/nio/channels/FileChannel/LargeGatheringWrite.java

[2022-02-24T05:21:01.277Z] STDERR:
[2022-02-24T05:21:01.277Z] java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
[2022-02-24T05:21:01.277Z] 	at java.base/java.nio.Bits.reserveMemory(Bits.java:178)
[2022-02-24T05:21:01.277Z] 	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123)
[2022-02-24T05:21:01.277Z] 	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:332)
[2022-02-24T05:21:01.277Z] 	at java.base/sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:243)
[2022-02-24T05:21:01.277Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:209)
[2022-02-24T05:21:01.277Z] 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:165)
[2022-02-24T05:21:01.277Z] 	at java.base/sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:317)
[2022-02-24T05:21:01.277Z] 	at java.base/java.nio.channels.FileChannel.write(FileChannel.java:436)
[2022-02-24T05:21:01.277Z] 	at LargeGatheringWrite.main(LargeGatheringWrite.java:91)
[2022-02-24T05:21:01.277Z] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
[2022-02-24T05:21:01.277Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
[2022-02-24T05:21:01.277Z] 	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
[2022-02-24T05:21:01.277Z] 	at java.base/java.lang.Thread.run(Thread.java:889)
[2022-02-24T05:21:01.277Z] 
[2022-02-24T05:21:01.277Z] JavaTest Message: Test threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
[2022-02-24T05:21:01.277Z] JavaTest Message: shutting down test
[2022-02-24T05:21:01.277Z] 
[2022-02-24T05:21:01.277Z] STATUS:Failed.`main' threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
```

**Comment by tajila** [comment]
The memory limit is determined by `j9gc_get_maximum_heap_size`. The new requested amount is 48bytes over the the limit

```
[2022-02-24T05:21:01.277Z] STATUS:Failed.`main' threw exception: java.lang.OutOfMemoryError: Cannot reserve 536870912 bytes of direct buffer memory (allocated: 3221225520, limit: 3758096384)
```

**Comment by tajila** [comment]
The test is run with Xmx4g, @dmitripivkine how is the max heap sized determined based on Xmx?

**Comment by tajila** [comment]
Also, has this changed recently?

**Comment by dmitripivkine** [comment]
`j9gc_get_maximum_heap_size()` returns back the exact value specified by `-Xmx` (use the same variable). I believe there were no changes recently in this area.

**Comment by tajila** [comment]
On xlinux

with Xmx4g
```
java.lang.Runtime.getRuntime().totalMemory() == 4g
jdk.internal.misc.VM.maxDirectMemory() == 4g
```

on windows

```
java.lang.Runtime.getRuntime().totalMemory() == 4g
jdk.internal.misc.VM.maxDirectMemory() == 3.7g
```

The cause is that the `sun.nio.MaxDirectMemorySize` property is set to 3.7g on windows but it doesnt appear to be set on xlinux

**Comment by tajila** [comment]
The tests above were done with openj9-jdk18

**Comment by pshipton** [comment]
There is code to set it at 7/8 of the heap size https://github.com/eclipse-openj9/openj9/blob/master/runtime/jcl/common/system.c#L476

On xlinux 18 I see the system property is set (with -Xmx4g) `sun.nio.MaxDirectMemorySize = 3758096384`

**Comment by pshipton** [comment]
With OpenJDK (8 - 18) I don't see any setting for MaxDirectMemorySize

**Comment by pshipton** [comment]
The OpenJ9 code comes from https://github.com/eclipse-openj9/openj9/pull/2629

**Comment by pshipton** [comment]
I don't see why OpenJ9 needs to set the limit at 7/8, seems it should just not set MaxDirectMemorySize at all.

**Comment by pshipton** [comment]
> The cause is that the sun.nio.MaxDirectMemorySize property is set to 3.7g on windows but it doesnt appear to be set on xlinux

This is weird because from the code it appears it's set the same on every platform, and I can see the property set on xlinux.

**Comment by tajila** [comment]
> This is weird because from the code it appears it's set the same on every platform, and I can see the property set on xlinux.

May mistake, I was using the RI JDK18. On j9 xlinux it is set as well

**Comment by pshipton** [comment]
We should validate that OpenJDK uses Runtime.getRuntime().totalMemory() as the default maxDirectMemory() for all versions, and if that is confirmed then remove setting sun.nio.MaxDirectMemorySize to be 7/8 of the max heap size, and just use the default. This doesn't need to occur for the Java 18 release, we can defer to 0.32.

**Comment by pshipton** [comment]
I'll keep this open until we decide if we are going to backport the fix.

**Comment by pshipton** [comment]
Fix is merged in head, 0.31, 0.32
